kernel/work_q.c

```diff
@@ -116,18 +114,18 @@ int k_delayed_work_cancel(struct k_delayed_work *work)
 {
 	int key = irq_lock();
 
-	if (k_work_pending(&work->work)) {
-		irq_unlock(key);
-		return -EINPROGRESS;
-	}
-
 	if (!work->work_q) {
 		irq_unlock(key);
 		return -EINVAL;
 	}
 
-	/* Abort timeout, if it has expired this will do nothing */
-	_abort_timeout(&work->timeout);
+	if (k_work_pending(&work->work)) {
+		/* Remove from the queue if already submitted */
+		k_queue_find_and_remove(&work->work_q->queue, work);
```

Comment:
The risk with doing this is that now the critical window between irq_lock() and irq_unlock() does not have a fixed bound anymore, since find_and_remove may need to potentially traverse the entire linked list. I'm not sure that's acceptable. We've already seen that too long irq_lock/unlock durations will mess up our BLE link layer which is almost purely low-latency interrupt based.